### PR TITLE
Embedded tests

### PR DESCRIPTION
**Pull request by ajmcmiddlin**
This should probably get some more rigorous testing before getting merged.

Provides an abstract representation of tests that can be run using either an embedded runner, or a runner backed by `tasty`. This allows students to run tests without any dependencies beyond `GHCi`.

**NOTE**: Property tests are only supported by the `tasty` runner

**NOTE**: Property tests that take generated functions are not currently supported at all.

### CODE REVIEW

**Comment by ajmcmiddlin** [comment]
@tonymorris 's comments

- Clarify the term 'embedded' in README
- `import qualified` test modules in `TestLoader` as is done in `TastyLoader`
   + Consistency
   + Ease of use

**Comment by ajmcmiddlin** [comment]
Some of the history has been cleaned up and it's been merged with master to resolve conflicts. Am doing some more thorough testing of things now. At the very least I think the newly added test modules will need porting.

**Comment by ajmcmiddlin** [comment]
I've now run through every part of the course that has tests and checked they work with both the built-in test runner and the tasty runner. I think this is good to merge.

File: test/Course/ListZipperTest.hs [hunk 1]
```diff
@@ -1,41 +1,84 @@
...
+import qualified Prelude            as P (fromIntegral, (<$>))
 
-import qualified Prelude                  as P (fromIntegral, (<$>))
-import           Test.QuickCheck.Function (Fun (..))
-import           Test.Tasty               (TestTree, testGroup)
-import           Test.Tasty.HUnit         (testCase, (@?=))
-import           Test.Tasty.QuickCheck    (testProperty)
+import           Course.Gens        (genInteger, genList, genListZipper,
+                                     genListZipperWithInt)
+import Test.Course.Mini (courseTest)
```
Comment (gwils) [inline test/Course/ListZipperTest.hs]: This isn't lined up with the others

File: test/Course/StateTest.hs [hunk 2]
```diff
@@ -1,35 +1,51 @@
...
+                                     firstRepeat, get, isHappy, put, runState)
 
-test_State :: TestTree
+test_State :: MiniTestTree
 test_State =
   testGroup "State" [
-    execTest
-  , evalTest
-  , getTest
+  --   execTest
```
Comment (gwils) [inline test/Course/StateTest.hs]: Is the commented-out code in this file meant to be here?

Comment (ajmcmiddlin) [reply test/Course/StateTest.hs]: These two are properties that require arbitrary functions, which aren't currently implemented for the built-in tests. I was leaving them commented so they weren't forgotten. Happy to leave an explanatory comment and get rid of the commented code. Thoughts?

File: test/Course/ListZipperTest.hs (hunk 1 above)
Comment (ajmcmiddlin) [reply test/Course/ListZipperTest.hs]: Thanks!

File: test/Course/JsonParserTest.hs [hunk 3]
```diff
@@ -2,11 +2,25 @@
...
+
+  -- * Course test runner
+  , courseTest
+  ) where
 
 import           Data.Ratio        ((%))
-import           Test.Tasty        (TestTree, testGroup)
-import           Test.Tasty.HUnit  (testCase, (@?=))
+import           Test.Course.Mini  (courseTest)
+import           Test.Mini         (MiniTestTree, Tester (..), UnitTester (..))
```
Comment (tonymorris) [inline test/Course/JsonParserTest.hs]: I think explicitly listed imports would help readers.

**Pull request merged**